src/cascadia/TerminalSettingsModel/Profile.cpp

```diff
@@ -169,50 +221,47 @@ winrt::com_ptr<winrt::Microsoft::Terminal::Settings::Model::implementation::Prof
 // - true iff the json object has the same `GUID` as we do.
 bool Profile::ShouldBeLayered(const Json::Value& json) const
 {
-    if (!_Guid.has_value())
-    {
-        return false;
-    }
-
     // First, check that GUIDs match. This is easy. If they don't match, they
     // should _definitely_ not layer.
-    if (const auto otherGuid{ JsonUtils::GetValueForKey<std::optional<winrt::guid>>(json, GuidKey) })
+    const auto otherGuid{ JsonUtils::GetValueForKey<std::optional<winrt::guid>>(json, GuidKey) };
+    const auto otherSource{ JsonUtils::GetValueForKey<std::optional<winrt::hstring>>(json, SourceKey) };
+    if (otherGuid)
     {
-        if (otherGuid.value() != *_Guid)
+        if (otherGuid.value() != Guid())
         {
             return false;
         }
     }
     else
     {
-        // If the other json object didn't have a GUID, we definitely don't want
-        // to layer. We technically might have the same name, and would
-        // auto-generate the same guid, but they should be treated as different
-        // profiles.
-        return false;
+        // If the other json object didn't have a GUID,
+        // check if we auto-generate the same guid using the name and source.
+        const auto otherName{ JsonUtils::GetValueForKey<std::optional<winrt::hstring>>(json, NameKey) };
+        if (Guid() != _GenerateGuidForProfile(otherName ? *otherName : L"Default", otherSource ? *otherSource : L""))
+        {
+            return false;
+        }
     }
 
-    std::optional<std::wstring> otherSource;
-    bool otherHadSource = JsonUtils::GetValueForKey(json, SourceKey, otherSource);
-
     // For profiles with a `source`, also check the `source` property.
     bool sourceMatches = false;
-    if (!_Source.empty())
+    const auto mySource{ Source() };
+    if (!mySource.empty())
     {
-        if (otherHadSource)
+        if (otherSource.has_value())
         {
             // If we have a source and the other has a source, compare them!
-            sourceMatches = *otherSource == _Source;
+            sourceMatches = *otherSource == mySource;
         }
         else
         {
             // Special case the legacy dynamic profiles here. In this case,
             // `this` is a dynamic profile with a source, and our _source is one
             // of the legacy DPG namespaces. We're looking to see if the other
             // json object has the same guid, but _no_ "source"
-            if (_Source == WslGeneratorNamespace ||
-                _Source == AzureGeneratorNamespace ||
-                _Source == PowershellCoreGeneratorNamespace)
+            if (mySource == WslGeneratorNamespace ||
+                mySource == AzureGeneratorNamespace ||
+                mySource == PowershellCoreGeneratorNamespace)
             {

```

Comment:
FYI @carlos-zamora  this is the code that makes "source" optional for the pre-0.5 dynamic generators